### PR TITLE
Don't break downloads when renewing

### PR DESCRIPTION
**Pull request by lukechampine**
This fixes a bug that caused files to become unretrievable after renewing or changing the allowance. It also fixes a deadlock in the renter.

### CODE REVIEW

File: modules/renter/renter.go [hunk 1]
```diff
@@ -177,6 +180,16 @@ func newRenter(cs modules.ConsensusSet, tpool modules.TransactionPool, hdb hostD
...
 	go r.threadedQueueRepairs()
+
+	// Kill workers on shutdown.
+	r.tg.OnStop(func() {
+		id := r.mu.RLock()
+		for _, worker := range r.workerPool {
+			close(worker.killChan)
+		}
+		r.mu.RUnlock(id)
+	})
```
Comment (DavidVorick) [inline modules/renter/renter.go]: This code is not needed, as each worker also listens for a tg.StopChan() signal.

But it does add clarity to have it here.

Comment (lukechampine) [inline modules/renter/renter.go]: I think there is actually a deadlock.
The worker loop looks like:
```go
for {
		select {
		case <-w.killChan:
			return
		default:
		}
		if w.renter.tg.Add() != nil {
			return
		}
		w.work()
		w.renter.tg.Done()
}
```
But `tg.Add` and `tg.Stop` both lock `tg.bmu`, so `tg.Add` will block until shutdown finishes.

**Comment by DavidVorick** [comment]
This needs a test demonstrating what was broken before + what is fixed now. You can probably set something up in the API using the ecosystem tests.

File: api/renterhost_test.go [hunk 2]
```diff
@@ -799,3 +799,93 @@ func TestRenterParallelDelete(t *testing.T) {
...
+		st.miner.AddBlock()
+	}
+	// Wait for the contract to be renewed.
+	time.Sleep(time.Second)
+
+	// Try downloading the file.
+	downpath := filepath.Join(st.dir, "testdown.dat")
+	err = st.stdGetAPI("/renter/download/test?destination=" + downpath)
+	if err != nil {
+		t.Fatal(err)
```
Comment (lukechampine) [inline api/renterhost_test.go]: needless to say, we should have written this test a while ago. It would have caught the bug while you were writing the worker code, instead of much later.

**Comment by DavidVorick** [comment]
Can I ask you to write additional tests around a few more edge cases?

1. Set the allowance to a little longer (so you have 10 blocks wiggle room instead of 5)
2. after setting the allowance, change it to have [change], then let it do it's thing, then try re-downloading
3. mine blocks until the period runs through again (an automatic one to follow the manual one you just did) and download again 
Changes:

{larger allowance, smaller allowance, same allowance}, {larger period, smaller period, same period}

That's 9 scenarios in total. Not sure if easy to make a table, but at the very least should should be able to just copy-paste between them and change the numbers around.
